mapiproxy/servers/default/emsmdb/emsmdbp_object.c

```diff
@@ -1234,6 +1234,68 @@ enum MAPISTATUS emsmdbp_folder_get_folder_count(struct emsmdbp_context *emsmdbp_
 	return retval;
 }
 
+static inline enum MAPISTATUS emsmdbp_object_move_folder_to_mapistore_root(struct emsmdbp_context *emsmdbp_ctx, struct emsmdbp_object *folder, struct emsmdbp_object *parent_folder, const char *new_name)
+{
+	uint64_t		parent_fid, fid, test_folder_id;
+	enum MAPISTATUS		retval;
+	char			*mapistore_uri;
+	TALLOC_CTX		*mem_ctx;
+
+
+	/* Sanity checks */
+	MAPI_RETVAL_IF(!emsmdbp_ctx, MAPI_E_INVALID_PARAMETER, NULL);
+	MAPI_RETVAL_IF(!folder, MAPI_E_INVALID_PARAMETER, NULL);
+	MAPI_RETVAL_IF(!parent_folder, MAPI_E_INVALID_PARAMETER, NULL);
+	MAPI_RETVAL_IF(!new_name, MAPI_E_INVALID_PARAMETER, NULL);
+
+	mem_ctx = talloc_zero(NULL, TALLOC_CTX);
+	MAPI_RETVAL_IF(!mem_ctx, MAPI_E_NOT_ENOUGH_MEMORY, NULL);
+
+	if (parent_folder->type == EMSMDBP_OBJECT_MAILBOX) {
+		parent_fid = parent_folder->object.mailbox->folderID;
+	}
+	else { /* EMSMDBP_OBJECT_FOLDER */
+		parent_fid = parent_folder->object.folder->folderID;
+	}
+
+	if (openchangedb_get_fid_by_name(emsmdbp_ctx->oc_ctx, emsmdbp_ctx->username, parent_fid,
+					 new_name, &test_folder_id) == MAPI_E_SUCCESS) {
+		DEBUG(4, ("emsmdbp_object: move_folder_to_mapistore_root duplicate folder error\n"));
+		retval = MAPI_E_COLLISION;
+		goto end;
+	}
+
+	fid = folder->object.folder->folderID;
+
+	retval = emsmdbp_get_uri_from_fid(mem_ctx, emsmdbp_ctx, fid, &mapistore_uri);
+        if (retval != MAPI_E_SUCCESS) {
+                DEBUG(0, ("Cannot locate folder id: %"PRIu64" on indexing database", fid));
+                goto end;
+        }
+
+	retval = openchangedb_create_folder(emsmdbp_ctx->oc_ctx, emsmdbp_ctx->username,
+					    parent_fid, fid, -1,
```

Comment:
you need to supply a change_number, actually I've spotted a bug thanks to your `-1`